documentation/design_docs/L2/tsop752.md

```diff
@@ -0,0 +1,131 @@
+
+For consumer IR applications, such as IR remote control systems, two of
+commonly used IR transmission protocol include RC-5 and NEC. An example of a
+typical IR transmission frame conforming to the NEC protocol can be found
+[here](https://techdocs.altium.com/sites/default/files/wiki_attachments/296329/NECMessageFrame.png).
+
+# Detailed Design
+## API
+```C++
+namespace sjsu
+{
+class Tsop752 final : public InfraredReceiver
+{
+ public:
+  constexpr explicit Tsop752(sjsu::Capture & capture,
+                             sjsu::Timer & timer);
+
+  Status Initialize() const override;
+  void SetInterruptHandler(DataReceivedHandler handler) const override;
+
+ private:
+  static void HandlePulseCaptured();
+  static void HandleEndOfDataFrame();
+
+  inline static DataFrame_t data_frame;
+  inline static DataReceivedHandler interrupt_handler;
+
+  const sjsu::Capture & capture_;
+  const sjsu::Timer & timer_;
+};
+}  // namespace sjsu
+```
+
+### Initialization
+```C++
+constexpr explicit Tsop752(sjsu::Capture & capture,
+                           sjsu::Timer & timer)
+
+Status Initialize() const override
+```
+The following sequence is performed to initialize the driver:
+1. Initialize and configure `capture_`.
+2. Initialize and configure `timer_`.
+3. Return `Status::kSuccess` if both drivers are initialize successfully,
+   Otherwise return the failure `Status`.
+
+### Capturing The De-modulated IR signal
+```C++
+void SetInterruptHandler(DataReceivedHandler handler) const override
+```
+`SetInterruptHandler` sets the `interrupt_handler` that is invoked when a data
+frame is successfully received.
+
+```C++
+static void HandlePulseCaptured()
+```
+The capture module is configured to increment at 1MHz (1µs).
+The `HandlePulseCaptured` is invoked when a pulse edge is detected and the
+value of Capture Register, CRn, will serve as a timestamp for when the edge was
+detected. The timestamp is appended to the `pulse_buffer` of `data_frame`.
+Additionally, upon capturing a pulse edge, the timer is reset.
+
+```C++
+static void HandleEndOfDataFrame()
+```
+The `HandleEndOfDataFrame` is invoked when the timeout period of the time is
+reached after receiving after a pulse edge is received. This signifies the end
+of the transmission frame. The `interrupt_handler` is invoked to allow the user
+to process the received data.
+
+# Caveats
+N/A
+
+# Future Advancements
+N/A
+
+# Testing Plan
+## Unit Testing Scheme
+The `Capture` and `Timer` modules shall be mocked and stubbed to provide test
+inputs for the `Tsop752` driver.
+
+|       Function      |    Expected Result    |
+|:--------------------|:----------------------|
+|Initialize           | Should return the failure `Status` `Initialize()` of either `Capture` or `Timer` are stubbed to not return `Status::kSuccess`.<br/> Should return `Status::kSuccess` when `Initialize()` of both `Capture` and `Timer` are stubbed to return `Status::kSuccess`. |
```

Comment:
Can you fix this? This looks really messed up. Try and keep it within 80 cols.

Reply:
Starting a new line would cause the sentence to appear in a new cell. Should I just use bullet points instead and stay away from tables?

Reply:
Yeah, use bullet points instead.